File: Raven.Client.Java/src/main/java/net/ravendb/client/document/batches/LazyStartsWithOperation.java
package net.ravendb.client.document.batches;

import java.util.ArrayList;
import java.util.List;

import net.ravendb.abstractions.data.Constants;
import net.ravendb.abstractions.data.GetRequest;
import net.ravendb.abstractions.data.GetResponse;
import net.ravendb.abstractions.data.JsonDocument;
import net.ravendb.abstractions.data.QueryResult;
import net.ravendb.abstractions.json.linq.RavenJArray;
import net.ravendb.abstractions.json.linq.RavenJObject;
import net.ravendb.client.RavenPagingInformation;
import net.ravendb.client.connection.IDatabaseCommands;
import net.ravendb.client.connection.SerializationHelper;
import net.ravendb.client.document.InMemoryDocumentSessionOperations;
import net.ravendb.client.utils.UrlUtils;


public class LazyStartsWithOperation<T> implements ILazyOperation {
  private final String keyPrefix;
  private final String matches;
  private final String exclude;
  private final int start;
  private final int pageSize;
  private final InMemoryDocumentSessionOperations sessionOperations;
  private final RavenPagingInformation pagingInformation;
  private final Class<T> clazz;
  private QueryResult queryResult;

  private Object result;
  private boolean requiresRetry;

  public LazyStartsWithOperation(Class<T> clazz, String keyPrefix, String matches, String exclude, int start, int pageSize, InMemoryDocumentSessionOperations sessionOperations, RavenPagingInformation pagingInformation) {
    this.clazz = clazz;
    this.keyPrefix = keyPrefix;
    this.matches = matches;
    this.exclude = exclude;
    this.start = start;
    this.pageSize = pageSize;
    this.sessionOperations = sessionOperations;
    this.pagingInformation = pagingInformation;
  }

  @Override
  public GetRequest createRequest() {
    int actualStart = start;
    boolean nextPage = pagingInformation != null && pagingInformation.isForPreviousPage(start, pageSize);
    if (nextPage) {
      actualStart = pagingInformation.getNextPageStart();
    }

    GetRequest getRequest = new GetRequest();
    getRequest.setUrl("/docs");
    getRequest.setQuery(String.format("startsWith=%s&matches=%s&exclude=%s&start=%d&pageSize=%d&next-page=%s", UrlUtils.escapeDataString(keyPrefix),
        UrlUtils.escapeDataString(matches != null ? matches : ""),
        UrlUtils.escapeDataString(exclude != null ? exclude : ""),
        actualStart, pageSize, nextPage ? "true" : "false"));
    return getRequest;
  }

  @Override
  public Object getResult() {
    return result;
  }

  @Override
  public QueryResult getQueryResult() {
    return queryResult;
  }


  public void setQueryResult(QueryResult queryResult) {
    this.queryResult = queryResult;
  }

  public void setResult(Object result) {
    this.result = result;
  }

  @Override
  public boolean isRequiresRetry() {
    return requiresRetry;
  }

  public void setRequiresRetry(boolean requiresRetry) {
    this.requiresRetry = requiresRetry;
  }

  @Override
  public void handleResponse(GetResponse response) {
    if (response.isRequestHasErrors()) {
      result = null;
      requiresRetry = false;
      return;
    }

    List<JsonDocument> jsonDocuments = SerializationHelper.ravenJObjectsToJsonDocuments(((RavenJArray)response.getResult()).values(RavenJObject.class));

    List<Object> resultList = new ArrayList<>();
    for (JsonDocument doc: jsonDocuments) {
      resultList.add(sessionOperations.trackEntity(clazz, doc));
    }

    if (pagingInformation !=null) {
      try {
        int nextPageStart = Integer.parseInt(response.getHeaders().get(Constants.NEXT_PAGE_START));
        pagingInformation.fill(start, pageSize, nextPageStart);
      } catch (NumberFormatException e) {
        // ignore
      }
    }

    result = resultList.toArray();
  }

  @Override
  public AutoCloseable enterContext() {
    return null;
  }

  public Object executeEmbedded(IDatabaseCommands commands) {
    List<JsonDocument> docs = commands.startsWith(keyPrefix, matches, start, pageSize);
    List<Object> entites = new ArrayList<>();
    for (JsonDocument doc: docs) {
      entites.add(sessionOperations.trackEntity(clazz, doc));
    }
    return entites.toArray();
  }

  public void handleEmbeddedResponse(Object result) {
    this.result = result;
  }


}

File: src/Raven.NewClient/Commands/Lazy/LazyStartsWithOperation.cs
// -----------------------------------------------------------------------
//  <copyright file="LazyStartsWithOperation.cs" company="Hibernating Rhinos LTD">
//      Copyright (c) Hibernating Rhinos LTD. All rights reserved.
//  </copyright>
// -----------------------------------------------------------------------

using System;
using Raven.NewClient.Client.Data;
using Raven.NewClient.Client.Document;
using Raven.NewClient.Client.Data.Queries;
using Raven.NewClient.Client.Document.Batches;
using Raven.NewClient.Client.Shard;
using Sparrow.Json;

namespace Raven.NewClient.Client.Commands.Lazy
{
   
    public class LazyStartsWithOperation<T> : ILazyOperation
    {
        private readonly string keyPrefix;

        private readonly string matches;

        private readonly string exclude;

        private readonly int start;

        private readonly int pageSize;

        private readonly InMemoryDocumentSessionOperations sessionOperations;

        private readonly RavenPagingInformation pagingInformation;
        private readonly string skipAfter;

        public LazyStartsWithOperation(string keyPrefix, string matches, string exclude, int start, int pageSize, InMemoryDocumentSessionOperations sessionOperations, RavenPagingInformation pagingInformation, string skipAfter)
        {
            this.keyPrefix = keyPrefix;
            this.matches = matches;
            this.exclude = exclude;
            this.start = start;
            this.pageSize = pageSize;
            this.sessionOperations = sessionOperations;
            this.pagingInformation = pagingInformation;
            this.skipAfter = skipAfter;
        }

        public GetRequest CreateRequest()
        {
            var actualStart = start;

            var nextPage = pagingInformation != null && pagingInformation.IsForPreviousPage(start, pageSize);
            if (nextPage)
                actualStart = pagingInformation.NextPageStart;

            return new GetRequest
            {
                Url = "/docs",
                Query =
                    string.Format(
                        "startsWith={0}&matches={3}&exclude={4}&start={1}&pageSize={2}&next-page={5}&skipAfter={6}",
                        Uri.EscapeDataString(keyPrefix),
                        actualStart,
                        pageSize,
                        Uri.EscapeDataString(matches ?? ""),
                        Uri.EscapeDataString(exclude ?? ""),
                        nextPage ? "true" : "false",
                        skipAfter)
            };
        }

        public object Result { get; set; }

        public QueryResult QueryResult { get; set; }

        public bool RequiresRetry { get; set; }

        public void HandleResponse(BlittableJsonReaderObject response)
        {
            throw new NotImplementedException();
            /*if (response.RequestHasErrors())
            {
                Result = null;
                RequiresRetry = false;
                return;
            }

            var jsonDocuments = SerializationHelper.RavenJObjectsToJsonDocuments(((RavenJArray)response.Result).OfType<RavenJObject>());

            int nextPageStart;
            if (pagingInformation != null && int.TryParse(response.Headers[Constants.Headers.NextPageStart], out nextPageStart))
                pagingInformation.Fill(start, pageSize, nextPageStart);

            Result = jsonDocuments
                .Select(sessionOperations.TrackEntity<T>)
                .ToArray();*/
        }
    }
}
